game-app/game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtySelector.java

```diff
@@ -150,13 +150,16 @@ public static CasualtyDetails selectCasualties(
       killAmphibiousFirst(killed, sortedTargetsToPickFrom);
     }
 
-    // if partial retreat isn't possible, then ensure units with marine bonus are removed first or
-    // last depending on whether the bonus is negative or positive
+    // if partial retreat isn't possible and it was an amphibious assault, then ensure units with
```

Comment:
You should specify here we are talking about partial retreat of land units (the absence of it meaning that land units either can retreat all together or cannot retreat). Retreating air only is a partial retreat too, and I assume whether or not the partial retreat of air only is possible is irrelevant to this matter.

Reply:
It's a good sign you're jumping in with feedback @Cernelius 

The context of this comment is very much the next 10 lines of code. For the more generic documentation, check for things that are in block comments `/** */` and the content in `docs`, comments starting with `//` are going to be about the nearby code.

I'll leave it to @trevan to consider incorporating this feedback further.

With that said, the commentary does help and the next block of code is quite heavy. I wonder if some explanation variables can help and make the commentary unnecessary (commentary is a source of maintenance, so we want it to be terse and concise).

What do you think of something like this @trevan ?
```
boolean amphibRetreatNotPossible = !Properties.getPartialAmphibiousRetreat(data.getProperties();
if(amphibRetreatNotPossible) {
  boolean marineCasualtiesSelected = casualtySelection.getKilled().stream()
                .anyMatch(
                    unit -> unit.getUnitAttachment().getIsMarine() != 0 && unit.getWasAmphibious());
   boolean marinesDamaged = casualtySelection.getDamaged().stream()
                .anyMatch(
                    unit ->
                        unit.getUnitAttachment().getIsMarine() != 0 && unit.getWasAmphibious()));
  if(marineCasualtiesSelected || marinesDamaged) {
     casualtySelection =
          casualtySelection.ensureUnitsWithPositiveMarineBonusAreTakenLast(sortedTargetsToPickFrom);
   }
}
```

It could also be worth considering to rename the various methods/lines to refer to amphibious(assaulting) units in case we have units that are amphibious but not marines (say something like amphibious vehicles).

Reply:
I removed the comments and added explanation variables.

By the way, a "marine" could easily be a tank, artillery, etc.  It just has to have the marine bonus set and it would need to have been offloaded by a sea transport before the battle.  So this code isn't specific to a unit called "marine".  It is solely looking at the attributes that are set in the unit definition (isMarine) and the status of being offloaded from a sea transport during the same combat phase (wasAmphibious).